src/liballoc/vec.rs

```diff
@@ -2000,7 +1999,83 @@ impl<T, I> SpecExtend<T, I> for Vec<T>
         <Vec<T> as SpecExtend<T, I>>::spec_extend(&mut vector, iterator);
         vector
     }
+}
+
+fn from_into_iter_source<T, I>(mut iterator: I) -> Vec<T>
+    where I: Iterator<Item=T> + InPlaceIterable + SourceIter<Source=IntoIter<T>>
+{
+    let mut insert_pos = 0;
+
+    // FIXME: how to drop values written into source when iteration panics?
+    //   tail already gets cleaned by IntoIter::drop
+    while let Some(item) = iterator.next() {
+        let source_iter = iterator.as_inner();
+        let src_buf = source_iter.buf.as_ptr();
+        let src_idx = source_iter.ptr;
+        unsafe {
+            let dst = src_buf.offset(insert_pos as isize);
+            debug_assert!(dst as *const _ < src_idx,
+                          "InPlaceIterable implementation produced more\
+                          items than it consumed from the source");
+            ptr::write(dst, item)
+        }
+        insert_pos += 1;
+    }
+
+    let src = iterator.as_inner();
+    let vec = unsafe { Vec::from_raw_parts(src.buf.as_ptr(), insert_pos, src.cap)};
+    mem::forget(iterator);
+    vec
+}
+
+impl<T> SpecFrom<T, IntoIter<T>> for Vec<T> {
+    fn from_iter(iterator: IntoIter<T>) -> Self {
+        // A common case is passing a vector into a function which immediately
+        // re-collects into a vector. We can short circuit this if the IntoIter
+        // has not been advanced at all.
+        if iterator.buf.as_ptr() as *const _ == iterator.ptr {
+            unsafe {
+                let vec = Vec::from_raw_parts(iterator.buf.as_ptr(),
+                                              iterator.len(),
+                                              iterator.cap);
+                mem::forget(iterator);
```

Comment:
cc @RalfJung

Reply:
Generally `ManuallyDrop` is preferred over `mem::forget`. Is that the question?
```rust
let iterator = ManuallyDrop::new(iterator);
let vec  = Vec::from_raw_parts(...);
```

Reply:
This is also an issue but not quite what I was concerned about.

Since both the source and the sink data reside in the same `IntoIter` allocation but `IntoIter::drop` only knows the source parts in its back I somehow need to manually handle dropping the data in the front if a panic occurs during `iterator.next()` calls

Reply:
This change hunk actually isn't new code, it's just code moved from `SpecExtend` to `SpecFrom`, but I can update this too.